### PR TITLE
Add support for miscellaneous opts

### PR DESCRIPTION
**Pull request by kmakiela**
This PR adds support for missing options in the following sections:
* general
* auth
* listen
* services

Additionally, a few changes had to be made in the existing code to accomodate for some of the new options.

### CODE REVIEW

File: test/config_parser_SUITE.erl [hunk 1]
```diff
@@ -33,6 +33,21 @@ equivalence(Config) ->
...
+    State1 = mongoose_config_parser_cfg:parse_file(CfgPath),
+    Hosts1 = mongoose_config_parser:state_to_host_opts(State1),
+    Opts1 = mongoose_config_parser:state_to_opts(State1),
+
+    TOMLPath = ejabberd_helper:data(Config, "miscellaneous.toml"),
+    State2 = mongoose_config_parser_toml:parse_file(TOMLPath),
+    Hosts2 = mongoose_config_parser:state_to_host_opts(State2),
+    Opts2 = mongoose_config_parser:state_to_opts(State2),
+
+    ?eq(Hosts1, Hosts2),
```
Comment (aleklisi) [inline test/config_parser_SUITE.erl:47]: I have mixed feelings about introducing shorter, less descriptive name for `?assertEqual`. Why not just use it?

File: test/config_parser_SUITE.erl [hunk 2]
```diff
@@ -33,6 +33,21 @@ equivalence(Config) ->
...
+    compare_unordered_lists(lists:filter(fun filter_config/1, Opts1), Opts2,
```
Comment (aleklisi) [inline test/config_parser_SUITE.erl:48]: Why do you need so many calls?
How about asserting them:
```suggestion
   ?assertEqual(lists:sort(Opts1), lists:sort(Opts2)).
```

File: test/config_parser_SUITE.erl [hunk 3]
```diff
@@ -62,6 +77,10 @@ compare_values({auth_method, _}, V1, V2) when is_atom(V1) ->
     ?eq([V1], V2);
 compare_values({s2s_addr, _}, {_, _, _, _} = IP1, IP2) ->
     ?eq(inet:ntoa(IP1), IP2);
+compare_values(services, V1, V2) ->
+    V1_m = proplists:get_value(service_mongoose_system_metrics, V1),
+    V2_m = proplists:get_value(service_mongoose_system_metrics, V2),
+    compare_unordered_lists(V1_m, V2_m);
```
Comment (aleklisi) [inline test/config_parser_SUITE.erl:83]: I understand that using single letter variables names is tempting but please avoid it as it is much harder for someone else to read.
```suggestion
compare_values(services, Val1, Val2) ->
    MIMSysMetrics1 = proplists:get_value(service_mongoose_system_metrics, Val1),
    MIMSysMetrics2 = proplists:get_value(service_mongoose_system_metrics, Val2),
    compare_unordered_lists(MIMSysMetrics1, MIMSysMetrics2);
```

File: src/config/mongoose_config_parser_toml.erl [hunk 4]
```diff
@@ -177,7 +205,9 @@ listener_opt(<<"service">>, <<"shaper_rule">>, V) -> [{shaper_rule, b2a(V)}];
 listener_opt(<<"service">>, <<"check_from">>, V) -> [{service_check_from, V}];
 listener_opt(<<"service">>, <<"hidden_components">>, V) -> [{hidden_components, V}];
 listener_opt(<<"service">>, <<"conflict_behaviour">>, V) -> [{conflict_behaviour, b2a(V)}];
-listener_opt(<<"service">>, <<"password">>, V) -> [{password, b2l(V)}].
+listener_opt(<<"service">>, <<"password">>, V) -> [{password, b2l(V)}];
+listener_opt(<<"service">>, <<"max_fsm_queue">>, V) -> [{max_fsm_queue, V}].
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_parser_toml.erl:209]: > Does it make sense to introduce a guard like:
> 
> or is it too defensive code?

No, don't do it, validation will be done separately.

File: src/config/mongoose_config_parser_toml.erl [hunk 5]
```diff
@@ -286,14 +318,20 @@ cyrsasl_external(M) -> {mod, b2a(M)}.
...
+                [#local_config{key = {K, Host}, value = V} || {K, V} <- OuterOpts]]
+            end.
 
 -spec is_inner_auth_opt(atom()) -> boolean().
 is_inner_auth_opt(auth_method) -> false;
 is_inner_auth_opt(allow_multiple_connections) -> false;
 is_inner_auth_opt(anonymous_protocol) -> false;
 is_inner_auth_opt(sasl_mechanisms) -> false;
+is_inner_auth_opt(extauth_instances) -> false;
 is_inner_auth_opt(_) -> true.
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_parser_toml.erl:335]: The previous version is much more readable. What is more, the suggested change introduces non-standard indentation. I very much prefer the simple and ASCII-art-free original version.

File: src/config/mongoose_config_parser_toml.erl [hunk 6]
```diff
@@ -522,6 +570,7 @@ process_host_item(M) ->
 %% path: host_config[].*
 -spec process_host_section(toml_key(), ejabberd:server(), toml_section()) -> config_list().
 process_host_section(<<"auth">>, Host, Content) ->
+    ct:pal("In host_config:~n~p~n~p", [Host, Content]),
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_parser_toml.erl:573]: Remove debug message.

File: src/config/mongoose_config_parser_toml.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (aleklisi) [reply src/config/mongoose_config_parser_toml.erl:335]: Let me fix the formatting:
```suggestion
-type not_inner_auth_opt() ::  auth_method | allow_multiple_connections | anonymous_protocol | sasl_mechanisms | extauth_instances.
                                                  
-spec is_inner_auth_opt(not_inner_auth_opt()) -> boolean().
is_inner_auth_opt(NotInnerAuthOpt) ->
    Opts = [auth_method, allow_multiple_connections, anonymous_protocol, sasl_mechanisms, extauth_instances],
    not lists:any(fun(Opt) -> Opt == NotInnerAuthOpt end, Opts).
```

File: src/config/mongoose_config_parser_toml.erl [hunk 7]
```diff
@@ -566,6 +615,26 @@ tls_cipher(#{<<"key_exchange">> := KEx,
     #{key_exchange => b2a(KEx), cipher => b2a(Cipher), mac => b2a(MAC), prf => b2a(PRF)};
 tls_cipher(Cipher) -> b2l(Cipher).
 
+set_overrides(State) ->
+    lists:foldl(fun(F, StateIn) -> F(StateIn) end, State, [
+        fun(S) -> maybe_override(S, override_global) end,
+        fun(S) -> maybe_override(S, override_local) end,
+        fun(S) -> maybe_override(S, override_acls) end
+        ]).
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_parser_toml.erl:623]: Honestly, the previous version looks much better for me, the readability would suffer a lot after the suggested change.

File: src/config/mongoose_config_parser_toml.erl [hunk 8]
```diff
@@ -111,7 +112,34 @@ process_general(<<"max_fsm_queue">>, V) ->
 process_general(<<"http_server_name">>, V) ->
     [#local_config{key = cowboy_server_name, value = b2l(V)}];
 process_general(<<"rdbms_server_type">>, V) ->
-    [#local_config{key = rdbms_server_type, value = b2a(V)}].
+    [#local_config{key = rdbms_server_type, value = b2a(V)}];
+process_general(<<"override_local">>, V) ->
+    [#local_config{key = override_local, value = V}];
+process_general(<<"override_global">>, V) ->
+    [#local_config{key = override_global, value = V}];
```
[the remark this answers is not in the document]
Comment (kmakiela) [reply src/config/mongoose_config_parser_toml.erl:119]: It might have been the `#config` record instead. Does not really matter since it is not in the final config anyway, these `override` options are stored separately in `#state` record. In TOML we don't have a global `State` while parsing so this is a workaround to later remove them from the config and add them to the appropriate `#state` field.

File: test/config_parser_SUITE.erl (hunk 1 above)
Comment (chrzaszcz) [reply test/config_parser_SUITE.erl:47]: I was the one who introduced the `eq` macro here like we do for some other suites. For me if used very frequently, it makes the code a bit easier to read, that's it. What is more, I do add a debug printout to it from time to time ;-)

File: test/config_parser_SUITE.erl (hunk 2 above)
Comment (chrzaszcz) [reply test/config_parser_SUITE.erl:48]: There are many corner cases where both config versions vary syntactically, but work the same. The complex functions are needed to handle them. And of course there can be nested unordered lists inside unordered lists.

File: test/config_parser_SUITE.erl (hunk 3 above)
Comment (chrzaszcz) [reply test/config_parser_SUITE.erl:83]: I agree that this module could be more readable after refactoring, but I also think that such refactoring should be done in a separate story as it would fall out of scope here.

File: src/config/mongoose_config_parser_toml.erl [hunk 9]
```diff
@@ -111,7 +112,34 @@ process_general(<<"max_fsm_queue">>, V) ->
...
+process_general(<<"override_acls">>, V) ->
+    [#local_config{key = override_acls, value = V}];
+process_general(<<"pgsql_users_number_estimate">>, V) ->
+    ?HOST_F([#local_config{key = {pgsql_users_number_estimate, Host}, value = V}]);
```
[the remark this answers is not in the document]
Comment (kmakiela) [reply src/config/mongoose_config_parser_toml.erl:123]: The options wrapped with `HOST_F` might be host-specific and the config expects to have an entry for such option for each host defined (and have this `Host` as a part of the key for `#local_config`). In this example, assuming we have 2 hosts: `host1` and `host2`, this part of the config should look like:
```
{local_config,{pgsql_users_number_estimate,<<"host1">>},V},
{local_config,{pgsql_users_number_estimate,<<"host2">>},V}
```

File: src/config/mongoose_config_parser_toml.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (kmakiela) [reply src/config/mongoose_config_parser_toml.erl:335]: To be honest I find the previous version more readable as well, I think it's pretty clear what it does at a first glance and there is no need for change here

File: src/config/mongoose_config_parser_toml.erl (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_parser_toml.erl:119]: Maybe we could introduce another temporary config type in this case to avoid the confusion? It could be even `{override, local | global | acls}` that would be there only if the value is `true` - otherwise return empty list.

File: src/config/mongoose_config_parser_toml.erl [hunk 10]
```diff
@@ -566,6 +615,26 @@ tls_cipher(#{<<"key_exchange">> := KEx,
...
 tls_cipher(Cipher) -> b2l(Cipher).
 
+set_overrides(State) ->
+    lists:foldl(fun(F, StateIn) -> F(StateIn) end, State, [
+        fun(S) -> maybe_override(S, override_global) end,
+        fun(S) -> maybe_override(S, override_local) end,
+        fun(S) -> maybe_override(S, override_acls) end
+        ]).
+
+maybe_override(S = {state, Config, _, _, _, _}, Name) ->
```
[the remark this answers is not in the document]
Comment (kmakiela) [reply src/config/mongoose_config_parser_toml.erl:625]: `#state` is a record defined in another module and from what I've read it is supposed to be operated on only by functions from that module. In other words, the `#state` record is simply unknown here. Regarding one-letter variable names I wanted to remain consistent with already written code. I will change that in a few places but I think having `V` as a shortcut for `Value` all over the place is not such a bad thing

File: src/config/mongoose_config_parser_toml.erl [hunk 11]
```diff
@@ -566,6 +615,26 @@ tls_cipher(#{<<"key_exchange">> := KEx,
...
+    case lists:keyfind(Name, 2, Config) of
+        {_, _, true} ->
+            S1 = mongoose_config_parser:Name(S),
+            C1 = lists:keydelete(Name, 2, Config),
+            mongoose_config_parser:set_opts(C1, S1);
+        {_, _, false} ->
```
[the remark this answers is not in the document]
Comment (kmakiela) [reply src/config/mongoose_config_parser_toml.erl:631]: You are right that `keyfind` returns a tuple or `false` atom. That is exactly why it is done this way, because if there is a `false` value in the config file i.e.
```
override_local = false
```
Then we will match on a tuple and we want to remove it from the config, without changing the `#state.override_local` value. If we don't match, there is no such tuple and no need for any action - that's why there is this distinction in the cases.

File: test/config_parser_SUITE.erl (hunk 1 above)
Comment (kmakiela) [reply test/config_parser_SUITE.erl:47]: I don't mind shorter names and `?eq` is pretty descriptive imo

File: test/config_parser_SUITE.erl (hunk 2 above)
Comment (kmakiela) [reply test/config_parser_SUITE.erl:48]: it's not quite that simple - there are nested structures in these lists and one call to `lists:sort` does not deal with all that

File: src/config/mongoose_config_parser_toml.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_parser_toml.erl:623]: I agree with @aleklisi that the function is a bit complicated. Let me try to find another solution. If, as I suggested above, options like `{override, local}` are used, we could just do

```suggestion
set_overrides(State) ->
    lists:foldl(fun({override, Type}) -> mongoose_config_parser:override(Type, State) end, State, Opts).
```
which would require a small change in `mongoose_config_parser` and its `cfg` backend.

**Comment by chrzaszcz** [comment]
Thanks @aleklisi for the review, it;s good to see another perspective, I added some comments as well. @kmakiela I would do another round after the code is merged/rebased from the current version of `toml-config`.

File: src/config/mongoose_config_parser_toml.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (kmakiela) [reply src/config/mongoose_config_parser_toml.erl:623]: Changed that, although it was not as simple because we still need to remove this `{override, Scope}`

File: src/config/mongoose_config_parser_toml.erl [hunk 12]
```diff
@@ -301,14 +338,20 @@ cyrsasl_external(M) -> {mod, b2a(M)}.
...
     {InnerOpts, OuterOpts} = lists:partition(fun({K, _}) -> is_inner_auth_opt(K) end, AuthOpts),
-    [#local_config{key = {auth_opts, Host}, value = InnerOpts} |
-     [#local_config{key = {K, Host}, value = V} || {K, V} <- OuterOpts]].
+    case InnerOpts of
+        [] ->
+            [#local_config{key = {K, Host}, value = V} || {K, V} <- OuterOpts];
+        _ ->
+            [#local_config{key = {auth_opts, Host}, value = InnerOpts} |
+                [#local_config{key = {K, Host}, value = V} || {K, V} <- OuterOpts]]
+            end.
```
Comment (chrzaszcz) [inline src/config/mongoose_config_parser_toml.erl:347]: I would **not** do this change now.

To give you the motivation, let me describe how the current version works:

- If there is an `auth` section in `host_config`, it overwrites all the globally set _inner_ options, changing them to defaults unless they are specified for the host.
- The advantage is that it is quite intuitive and in line with the `cfg` logic: a section specified for the host overwrites the default section.
- The disadvantage is that only the inner option are overwritten with defaults.

The version proposed in this PR would work like this:

- If there is any of the _inner_ options in the `auth` section for a host, it overwrites all the _inner_ options, changing them to defaults unless they are specified for the host.
- The disadvantage is that it is counter-intuitive as despite having an `auth` section defined, it would overwrite the global settings only if a specific option is present.

As I do not see the proposed changes as a solution and the big tests also depend on the current state, I would not do this change now.

However, I think it needs to be resolved in such a way that the distinction between _inner_ and _outer_ options disappears completely and we need to do this when we drop the support for the `cfg` format.

I hope this clarifies my view.

File: src/config/mongoose_config_parser_toml.erl [hunk 13]
```diff
@@ -599,6 +652,18 @@ tls_cipher(#{<<"key_exchange">> := KEx,
     #{key_exchange => b2a(KEx), cipher => b2a(Cipher), mac => b2a(MAC), prf => b2a(PRF)};
 tls_cipher(Cipher) -> b2l(Cipher).
 
+set_overrides(State) ->
+    Opts = mongoose_config_parser:get_opts(State),
+    lists:foldl(fun maybe_override/2, State, Opts).
+
+maybe_override(Override = {override, Type}, State) ->
+    Opts = mongoose_config_parser:get_opts(State),
+    NewOpts = lists:delete(Override, Opts),
```
Comment (chrzaszcz) [inline src/config/mongoose_config_parser_toml.erl:661]: I think we could do this before this step.

**Comment by kmakiela** [comment]
Thanks for implementing the remaining changes @chrzaszcz, looks good to me

**Review by aleklisi — APPROVED** [review]
